### PR TITLE
feat(resource): ask for metrics only when needed

### PR DESCRIPTION
**Pull request by keynslug**
Followup to [EMQX-9136](https://emqx.atlassian.net/browse/EMQX-9136), [EMQX-8375](https://emqx.atlassian.net/browse/EMQX-8375).

Followup to https://github.com/emqx/emqx/pull/10123#discussion_r1138285457.

### <samp>🤖 Generated by Copilot at a71ce2e</samp>

This pull request refactors the resource metrics management in the `emqx_resource` app and introduces a new version of the bridge protocol, `emqx_bridge_proto_v4`, in the `emqx_bridge` app. The new protocol improves the efficiency and scalability of bridge metrics retrieval and separates them from the bridge status and configuration data. The pull request also updates the `emqx_authn` and `emqx_authz` apps to use the new protocol and modifies some test cases accordingly.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [x] Changed lines covered in coverage report
- [x] Change log has been added to `changes/{ce,ee}/(feat|perf|fix)-<PR-id>.en.md` files
- [ ] ~~For internal contributor: there is a jira ticket to track this change~~
- [ ] ~~If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up~~
- [x] Schema changes are backward compatible

### CODE REVIEW

File: changes/ce/feat-10359.en.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Metrics now are not implicitly collected in places where API handlers does not make any use of them. Instead, a separate BPAPI RPC gathers cluster-wide metrics.
```
Comment (ieQu1) [inline changes/ce/feat-10359.en.md:1]:
```suggestion
Metrics now are not implicitly collected in places where API handlers don't make any use of them. Instead, a separate BPAPI RPC gathers cluster-wide metrics.
```

File: apps/emqx_bridge/src/emqx_bridge_api.erl [hunk 2]
```diff
@@ -990,7 +988,7 @@ do_bpapi_call(Node, Call, Args) ->
 do_bpapi_call_vsn(SupportedVersion, Call, Args) ->
     case lists:member(SupportedVersion, supported_versions(Call)) of
         true ->
-            apply(emqx_bridge_proto_v3, Call, Args);
+            apply(emqx_bridge_proto_v4, Call, Args);
```
Comment (savonarola) [inline apps/emqx_bridge/src/emqx_bridge_api.erl:991]: Nit: isn't this a kind of evading from some of the bpapi checks? 🤔

File: apps/emqx_resource/src/emqx_resource_manager.erl [hunk 3]
```diff
@@ -37,7 +37,6 @@
     list_all/0,
```
Comment (thalesmg) [inline apps/emqx_resource/src/emqx_resource_manager.erl:37]: Just curiosity: `list_all/0` still collects the metrics, and it seems to be used only by `emqx_resource:list_instances_verbose`, which in turn seems to be for debugging purposes.  Maybe we can remove `data_record_to_external_map_with_metrics`?

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_bridge/src/emqx_bridge_api.erl (hunk 2 above)
Comment (keynslug) [reply apps/emqx_bridge/src/emqx_bridge_api.erl:991]: What scenario do you have in mind? I believe that piece of logic always (silently) assumed that the type (i.e. signature) of an existing RPC does not change across protocol versions. Which is general is not true, but I'm not aware of bpapi checks for that.

Comment (savonarola) [reply apps/emqx_bridge/src/emqx_bridge_api.erl:991]: I mean, dialyzer won't check that `Args` type matches the bpapi's wrapper type; only that bpapi's wrapper declared type matches the type of the corresponding remote call.

**Review by savonarola — DISMISSED** [review]

**Comment by savonarola** [comment]
Things got much more readable when they got separated.

Do I understand correctly that cluster calls in bridge API ware needed only for metrics?

File: apps/emqx_resource/src/emqx_resource_manager.erl (hunk 3 above)
Comment (thalesmg) [reply apps/emqx_resource/src/emqx_resource_manager.erl:37]: I think the latter would be better, so we could drop the `..._map_with_metrics` fn and use the same API to collect metrics when calling this debug function.  What do you think?

Comment (keynslug) [reply apps/emqx_resource/src/emqx_resource_manager.erl:37]: I think mostly the same, did that in 9c9f39d0f7431ae7f17b711ff61c17135a223761.

**Review by savonarola — APPROVED** [review]

**Pull request merged**